### PR TITLE
Start a chapter on 3.11 interpreter internals

### PR DESCRIPTION
**Pull request by gvanrossum**
I haven't been able to finish this (see TODOs at the end), but I figure that maybe someone else might be interested, or at some point in the future I'd get another bit of inspiration. In the meantime, I'd rather not have my effort so far go to a complete waste.

Note that the 3.12 details are often quite different and deserve a separate chapter, but as 3.12 is still very much in flux I'd rather wait with that.

Who else should I ask for a review? Maybe @brandtbucher ?

### CODE REVIEW

File: internals/interpreter.rst [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+(Details differ depending on the ``is_entry`` flag.)
+When the generator is resumed, the interpreter pushes the ``_PyInterpreterFrame`` onto the stack and resumes execution.
+(There is more hairiness for generators and their ilk, we'll discuss these in a later section in more detail.)
+
+(TODO: Also frame layout and use, and "locals plus".)
+
+All sorts of variables
+======================
+
+The bytecode compiler determines for each variable name in which scope it is defined and generates instructions accordingly.
```
Comment (CAM-Gerlach) [inline internals/interpreter.rst:280]:
```suggestion
The bytecode compiler determines the scope in which each variable name is defined, and generates instructions accordingly.
```

Simplified the original's phrasing and tried to make it much more readible—I had a really hard time parsing it previously.

File: internals/interpreter.rst [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+For example, loading a local variable onto the stack is done using ``LOAD_FAST``, while loading a global is done using ``LOAD_GLOBAL``.
+The key types of variables are:
+- fast locals: used in functions
```
Comment (CAM-Gerlach) [inline internals/interpreter.rst:283]:
```suggestion
The key types of variables are:

- fast locals: used in functions
```

The line break is syntactically required here to make this a list

File: internals/interpreter.rst [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+- (slow or regular) locals: used in classes and at the top level
+- globals and builtins: the compiler does not distinguish between globals and builtins (though the specializing interpreter does)
+- cells -- used for nonlocal references
```
Comment (CAM-Gerlach) [inline internals/interpreter.rst:286]:
```suggestion
- cells: used for nonlocal references
```

To be consistent with the other three

File: internals/interpreter.rst [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+- A relative jump forward (``JUMP_FORWARD``) sets ``next_instr += oparg``.
+- A relative jump backward sets ``next_instr -= oparg``.
+
+A relative jump whose ``oparg`` is zero is a no-op.
+
+Inline cache entries
+====================
+
+Some (specialized or specializable) instructions have an associated "inline cache".
+The inline cache consists of one or more two-byte entries included in the bytecode array.
```
[the remark this answers is not in the document]
Comment (gvanrossum) [reply internals/interpreter.rst:119]: Done now.

File: internals/interpreter.rst [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+
+Serializing non-zero cache entries would present a problem because the serialization (``marshal``) format must be independent of the machine byte order.
+
+More information about the use of inline caches can be found in :pep:`659` (search for "ancillary data").
+
+The evaluation stack
+====================
+
+Apart from unconditional jumps, almost all instructions read or write some data in the form of object references (``PyObject *``).
+The CPython 3.11 bytecode interpreter is a stack machine, meaning that it operates by pushing data onto and popping it off the stack.
```
[the remark this answers is not in the document]
Comment (gvanrossum) [reply internals/interpreter.rst:141]: I'm adding "The stack is a pre-allocated array of object references." which should make the connection a bit more directly.

File: internals/interpreter.rst [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+In 3.10 and before this was the case even when a Python function called another Python function:
+The ``CALL`` instruction would call the ``tp_call`` dispatch function of the callee, which would extract the code object, create a new frame for the call stack, and then call back into the interpreter.
+This approach is very general but consumes several C stack frames for each nested Python call, thereby increasing the risk of an (unrecoverable) C stack overflow.
+
+In 3.11 the ``CALL`` instruction special-cases function objects to "inline" the call.
+When a call gets inlined, a new frame gets pushed onto the call stack and the interpreter "jumps" to the start of the callee's bytecode.
+When the callee executes a ``RETURN_VALUE`` instruction, the frame is popped off the call stack and the interpreter returns to the caller.
+There is a flag in the frame (``frame->is_entry``) that indicates whether the frame was inlined (set if it wasn't).
+If ``RETURN_VALUE`` returns to a caller where this flag is set, it performs the usual cleanup and return from ``_PyEval_EvalFrameDefault()``.
```
[the remark this answers is not in the document]
Comment (gvanrossum) [reply internals/interpreter.rst:248]: I've updated the paragraph.

**Comment by gvanrossum** [comment]
I've just landed this, I need to move on to 3.12 again. :-) (https://github.com/faster-cpython/ideas/pull/539, https://github.com/faster-cpython/ideas/issues/540).

File: internals/interpreter.rst [hunk 7]
```diff
@@ -0,0 +1,301 @@
+.. _interpreter:
+
+=====================================
+The CPython 3.11 Bytecode Interpreter
+=====================================
+
+.. highlight:: none
```
[the remark this answers is not in the document]
Comment (CAM-Gerlach) [reply internals/interpreter.rst]: I kept this as it was so it wouldn't try to mark up the (presumably plain text) lists of opcodes, and instead marked up the C blocks as such, but it looks like it worked out okay in the rendered result.

File: internals/interpreter.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (CAM-Gerlach) [reply internals/interpreter.rst]: It took a another re-read or two, but I think I've got it now with the revised version, thanks :)

File: internals/interpreter.rst [hunk 8]
```diff
@@ -0,0 +1,301 @@
...
+
+Instruction decoding
+====================
+
+The first task of the interpreter is to decode the bytecode instructions.
+Bytecode is stored as an array of 16-bit code units (``_Py_CODEUNIT``).
+Each code unit contains an 8-bit ``opcode`` and an 8-bit argument (``oparg``), both unsigned.
+In order to make the bytecode format independent of the machine byte order when stored on disk, ``opcode`` is always the first byte and ``oparg`` is always the second byte.
+Macros are used to extract the ``opcode`` and ``oparg`` from a code unit (``_Py_OPCODE(word)`` and ``_Py_OPARG(word)``).
+Some instructions (e.g. ``NOP`` or ``POP_TOP``) have no argument -- in this case we ignore ``oparg``.
```
[the remark this answers is not in the document]
Comment (CAM-Gerlach) [reply internals/interpreter.rst:62]: Yeah, this is a pretty fine-toothed typographical nitpick that most people would never notice; I just got it drilled into me young as my mom was a graphic designer

File: internals/interpreter.rst [hunk 9]
```diff
@@ -0,0 +1,301 @@
...
+The CPython 3.11 bytecode interpreter (a.k.a. virtual machine) has a number of improvements over 3.10.
+We describe the inner workings of the 3.11 interpreter here, with an emphasis on understanding not just the code but its design.
+While the interpreter is forever evolving, and the 3.12 design will undoubtedly be different again, understanding the 3.11 design will help you understand future improvements to the interpreter.
+
+Introduction
+============
+
+The bytecode interpreter's job is to execute Python code.
+Its main input is a code object, although this is not a direct argument to the interpreter.
+The interpreter is structured as a (recursive) function taking a thread state (``tstate``) and a stack frame (``frame``).
```
[the remark this answers is not in the document]
Comment (CAM-Gerlach) [reply internals/interpreter.rst:21]: Oops, I meant to delete this comment, as I made it before I later added a suggestion to link to ceval at the top (as at that point, I wasn't even 100% sure what file it was).

File: internals/interpreter.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (CAM-Gerlach) [reply internals/interpreter.rst]: Thanks!

File: internals/interpreter.rst (hunk 5 above)
[the remark this answers is not in the document]
Comment (CAM-Gerlach) [reply internals/interpreter.rst:141]: Yeah, still not perfect but better, I think :+1:

**Comment by hugovk** [comment]
The lesser-spotted ["explanation" part of Diátaxis](https://diataxis.fr/explanation/) :)

**Comment by CAM-Gerlach** [comment]
There actually is a [todo extensionsion](https://www.sphinx-doc.org/en/master/usage/extensions/todo.html) built into Sphinx, which has some handy features to mark todos, list them and control whether and how they are shown in the output, but I didn't want to complicate your life more for now by suggesting you use it :)

**Pull request merged**